### ISSUE TITLE
It would be really nice to have a collection of test data to test client conformance

### ISSUE THREAD

**Issue by patflynn** [theupdateframework/specification#260]
Generating test data for TUF clients is time consuming and error prone.  It could be really cool to have a test data set to exercise clients and certify that they're implementing all elements of the update algorithm.

**Comment by znewman01** [comment]
Glad you filed this, it comes up over and over and AFAIK there's no place to track.

There's at least [one abortive attempt](https://github.com/advancedtelematic/tuf-test-vectors) at this along with [semi-related prior art](https://theupdateframework.github.io/python-tuf/2022/06/15/testing-ngclient.html). It also comes up pretty often in the [go-tuf meeting](https://hackmd.io/dm0C3LpDTHOsQYZfUXRcBw) (see 2022-08-03 meeting, for instance) and the [TUF community meeting](https://hackmd.io/jdAk9rmPSpOYUdstbIvbjw) (see 2022-07-27, 2021-08-25 meetings).

There are two related suites of tests:

1. Conformance tests: we should be able to write these based solely on the spec
2. Interoperability tests: each client implementation should be able to consume the output of each other repository implementation (test matrix)

There are a couple of issues that make this nontrivial (but still very worth doing!):

- The different libraries present very diverse APIs, and don't agree on e.g. the primitive operations that a client would do. It's hard to (doubly so for the interop tests, when we want each client to generate a repo according to a template).
  - This makes it hard to create e.g. a YAML doc that describes various test inputs/outputs.
- There's a fair amount of flexibility permitted in the client implementations. For instance, the [POUF system](https://theupdateframework.github.io/specification/latest/index.html#pouf-documents) means that, strictly speaking, a client implementation doesn't need to understand JSON to be conformant. 
- There's a huge variety of scenarios to test. It might be worthwhile to use a fuzzing approach.

**Comment by asraa** [comment]
Yes!! We were talking about this together in the Sigstore world and definitely think it's useful to start with the Sigstore TUF clients. There's a starter [conformence tests](https://github.com/trailofbits/sigstore-conformance) for Sigstore clients that we can piggy-back off of for the TUF root.

* Diverse APIs: We can probably dictate with Sigstore TUF clients on their API
* All sigstore tuf client implementations MUST be able to understand the Sigstore TUF repository
* We definitely should catch corner cases as we update the Sigstore TUF client for more data

**Comment by joshuagl** [comment]
Zack already linked to the description of how [we handle testing in python-tuf](https://theupdateframework.github.io/python-tuf/2022/06/15/testing-ngclient.html). I also want to point out the [repository-editor-for-tuf](https://github.com/vmware-labs/repository-editor-for-tuf), which we've found to be a useful tool for generating metadata and reviewing the changes to the metadata over time/TUF operations (because repository-editor-for-tuf uses a git repository for storage).

**Comment by patflynn** [comment]
@joshuagl that repository editor tool looks fantastic thanks for the pointer. 

I think the differences in API aren't a huge deal. If we have a solid suite of test data it should be trivial for each client to implement a test driver for their API.

**Comment by znewman01** [comment]
> I think the differences in API aren't a huge deal. If we have a solid suite of test data it should be trivial for each client to implement a test driver for their API.

I wish I agreed with you :(

Some of the things the libraries disagree about include whether it's possible to point at an on-disk/in-memory TUF repo, whether it's possible to iterate targets metadata, whether you can access metadata at all or you just get a "download file" primitive, etc.

I think/hope in the medium term that there will be some convergence here, and perhaps such a test suite can drive convergence.
